lib/trino-parquet/src/main/java/io/trino/parquet/reader/LongColumnReader.java

```diff
@@ -28,6 +29,17 @@ public LongColumnReader(RichColumnDescriptor descriptor)
     @Override
     protected void readValue(BlockBuilder blockBuilder, Type type)
     {
-        type.writeLong(blockBuilder, valuesReader.readLong());
+        long longValue = valuesReader.readLong();
+
+        if (type instanceof TimestampType) {
+            int precision = ((TimestampType) type).getPrecision();
+            if (precision == 3) {
+                // When the 'timestamp(3)' (millisecond precision) is read from the column,
+                // just make sure it is converted to microsecond precision, which TimestampType expects
+                longValue *= 1000;
```

Comment:
What does the Hive do in such a case?

Add product tests covering this, and proving we are consistent.

Reply:
Then, why are we thinking of adding this non-standard (per Hive behavior) conversion?

Reply:
Just presenting a way how it *can be* handled. It does seem like a "quick win" that many users might find valuable, or at least a way to address possible nasty bug (see https://github.com/trinodb/trino/pull/10661#issuecomment-1016457070).